src/query/storage/m3/storage.go

```diff
@@ -257,20 +260,28 @@ func (s *m3storage) Write(
 	id := query.Tags.ID()
 	// TODO: Consider caching id -> identID
 	identID := ident.StringID(id)
-	common := &writeRequestCommon{
-		store:       s,
-		annotation:  query.Annotation,
-		unit:        query.Unit,
-		id:          identID,
-		tagIterator: storage.TagsToIdentTagIterator(query.Tags),
-		attributes:  query.Attributes,
-	}
+	tagIterator := storage.TagsToIdentTagIterator(query.Tags)
+
+	var (
+		wg       sync.WaitGroup
+		multiErr xerrors.MultiError
+	)
+
+	for _, datapoint := range query.Datapoints {
+		s.writeWorkerPool.Go(func() {
+			wg.Add(1)
+			tagIter := tagIterator.Duplicate()
```

Comment:
Are these threadsafe? Also since these all use the same iterator, can you define this outside of the loop?

Reply:
yeah makes sense